File: frontend/src/api/comfyui.ts
import { apiGet, apiPost, apiUrl } from "./client";
import { comfyClientId } from "../lib/comfyProgress";
import type { RegenerationSnapshot } from "../types/chat";

export interface ComfyStatus {
  running: boolean;
  managed: boolean;
}

export function comfyStatus(url: string) {
  // 6s 超时：NodeCard 等组件用它做 5s 轮询探活。后端 is_up 最坏可达 ~10s（HTTP 5s + TCP 5s），
  // 且 8010 繁忙时响应更慢——若无超时，单次请求挂起会让调用方的轮询链永久中断（表现为
  // 「ComfyUI 未启动，等待中…」后不再加载）。
  return apiGet<ComfyStatus>(`/comfyui/status?url=${encodeURIComponent(url)}`, 6000);
}

export function startComfy(path: string, url: string, pythonPath = "") {
  return apiPost<{ running: boolean; managed: boolean; message: string }>("/comfyui/start", {
    path,
    url,
    python_path: pythonPath,
  });
}

// 关闭 ComfyUI（装插件/依赖前需先关）
export function stopComfy(url: string, path = "") {
  return apiPost<{ stopped: boolean; message: string }>("/comfyui/stop", { url, path });
}

// 重启 ComfyUI（装完插件生效）：先关再起，需 path 重新拉起
export function restartComfy(path: string, url: string, pythonPath = "") {
  return apiPost<{ running: boolean; managed: boolean; message: string }>("/comfyui/restart", {
    path,
    url,
    python_path: pythonPath,
  });
}

// 把 ComfyUI 路径/地址落盘到后端，供 start-dev 脚本读取
export function saveComfyConfig(path: string, url: string, pythonPath = "") {
  return apiPost<{ path: string; url: string; python_path: string }>("/comfyui/config", {
    path, url, python_path: pythonPath,
  });
}

export interface SubmitResult {
  ok: boolean;
  prompt_id?: string;
  node_count?: number;
}

export function submitWorkflow(
  templateId: string,
  values: Record<string, unknown>,
  url: string,
  prompt = "",
  loras: { name: string; weight: number }[] = [],
  loraMode: "none" | "single" | "multi" = "single",
) {
  return apiPost<SubmitResult>("/comfyui/submit", {
    template_id: templateId,
    values,
    prompt,
    url,
    client_id: comfyClientId(),
    loras,
    lora_mode: loraMode,
  });
}

export interface UploadResult {
  name: string;
  raw: Record<string, unknown>;
}

// 上传图片到 ComfyUI 的 input 目录，返回可供 LoadImage 引用的文件名
export async function uploadImage(file: File, url: string): Promise<UploadResult> {
  const fd = new FormData();
  fd.append("file", file);
  fd.append("url", url);
  const resp = await fetch(apiUrl("/comfyui/upload"), { method: "POST", body: fd });
  if (!resp.ok) throw new Error(`API request failed: ${resp.status}`);
  return resp.json();
}

export interface ResultImage {
  filename: string;
  subfolder: string;
  type: string;
}

export interface GenResult {
  status: "pending" | "running" | "completed" | "failed" | "not_found";
  error?: string;
  images: ResultImage[];
  videos: ResultImage[];
  audios: ResultImage[];
  texts: string[];
}

export function getResult(promptId: string, url: string, nodeIds: string[] = []) {
  const extra = nodeIds.length > 0 ? `&node_ids=${encodeURIComponent(nodeIds.join(","))}` : "";
  return apiGet<GenResult>(
    `/comfyui/result?prompt_id=${encodeURIComponent(promptId)}&url=${encodeURIComponent(url)}${extra}`,
  );
}

export interface FinalizedMessage {
  id: string;
  role: "assistant";
  text: string;
  image?: string;
  video?: string;
  audio?: string;
  regeneration?: RegenerationSnapshot;
}

export interface FinalizeGenerationResponse {
  prompt_id: string;
  durable: boolean;
  complete: boolean;
  messages: FinalizedMessage[];
  images: { message_id: string; display_url: string; persisted: boolean; indexed: boolean; snapshotted: boolean; errors: string[] }[];
  target?: { message_id: string; slot_id: string; media_type: "image" | "video" | "audio"; url: string } | null;
}

export function finalizeGeneration(args: {
  threadId: string; repoId: string; promptId: string; prompt: string;
  images: ResultImage[]; videos?: ResultImage[]; audios?: ResultImage[]; outputDir: string; comfyuiUrl: string;
  embed: { baseUrl: string; apiKey: string; modelName: string };
  chat: { baseUrl: string; apiKey: string; modelName: string };
  regeneration?: RegenerationSnapshot;
  templateName?: string; modelName?: string; loraNames?: string[];
  target?: { messageId: string; slotId: string };
  baseSlotRef?: { messageId: string; slotId: string };
}) {
  return apiPost<FinalizeGenerationResponse>("/comfyui/finalize-generation", {
    thread_id: args.threadId, repo_id: args.repoId, prompt_id: args.promptId,
    prompt: args.prompt, images: args.images, videos: args.videos || [], audios: args.audios || [], output_dir: args.outputDir,
    comfyui_url: args.comfyuiUrl, embed_base: args.embed.baseUrl,
    embed_key: args.embed.apiKey, embed_model: args.embed.modelName,
    chat_base: args.chat.baseUrl, chat_key: args.chat.apiKey, chat_model: args.chat.modelName,
    regeneration: args.regeneration,
    template_name: args.templateName || "",
    model_name: args.modelName || "",
    lora_names: args.loraNames ? args.loraNames.join(",") : "",
    target_message_id: args.target?.messageId || "",
    target_slot_id: args.target?.slotId || "",
    base_slot_ref: args.baseSlotRef || null,
  });
}

// 音频分条按顺序拼接成完整版（后端 ffmpeg concat + 落盘回写快照）
export function mergeAudio(args: { threadId: string; messageId: string; force?: boolean }) {
  return apiPost<{ ok: boolean; url: string }>("/comfyui/merge-audio", {
    thread_id: args.threadId,
    message_id: args.messageId,
    force: !!args.force,
  });
}

// 拼出经后端代理的取图地址
export function viewUrl(img: ResultImage, url: string): string {
  const qs = new URLSearchParams({
    filename: img.filename,
    type: img.type,
    subfolder: img.subfolder,
    url,
  });
  return apiUrl(`/comfyui/view?${qs.toString()}`);
}

// P5 首尾帧顺序链：把 ComfyUI 生成的产物图（output 目录）转成视频模板可引用的
// input 文件名（fetch 取回 → 上传回 input 目录）。视频模板的 first_frame_image/
// last_frame_image 期望 LoadImage 可引用的 input 文件名（B3 同套路）。
export async function moveComfyOutputToInput(img: ResultImage, url: string): Promise<string> {
  const blob = await (await fetch(viewUrl(img, url))).blob();
  const file = new File([blob], img.filename, { type: blob.type || "image/png" });
  return (await uploadImage(file, url)).name;
}

// W3 转场视频：把「上尾帧图」（任意可 fetch 的 URL，如 localViewUrl 包装后的本地留存路径）
// 取回并上传回 ComfyUI input 目录，供转场视频模板的 first_frame_image（图片1=起点）引用。
export async function uploadRemoteImageToInput(srcUrl: string, comfyuiUrl: string): Promise<string> {
  const resp = await fetch(srcUrl);
  if (!resp.ok) throw new Error(`获取转场参考图失败：HTTP ${resp.status}`);
  const blob = await resp.blob();
  const name = srcUrl.split(/[\\/]/).pop() || "ref.png";
  const file = new File([blob], name, { type: blob.type || "image/png" });
  return (await uploadImage(file, comfyuiUrl)).name;
}

// 把原图留存到设置的 outputDir，返回本地文件路径
export function saveLocal(args: {
  img: ResultImage; repoId: string; outputDir: string; url: string;
}) {
  return apiPost<{ ok: boolean; path: string }>("/comfyui/save-local", {
    filename: args.img.filename,
    subfolder: args.img.subfolder,
    type: args.img.type,
    repo_id: args.repoId,
    output_dir: args.outputDir,
    url: args.url,
  });
}

// 本地留存原图的访问地址
export function localViewUrl(path: string): string {
  return apiUrl(`/comfyui/local-view?path=${encodeURIComponent(path)}`);
}

// 通用模式留存：把任意图片地址（云端直链 / data URI）存到 outputDir。
// subdir 非空时落到 <repo>/<subdir>/ 子夹（用户上传的参考图 → reference）。
export function saveLocalSrc(args: { src: string; repoId: string; outputDir: string; subdir?: string }) {
  return apiPost<{ ok: boolean; path: string }>("/comfyui/save-local", {
    src: args.src,
    repo_id: args.repoId,
    output_dir: args.outputDir,
    subdir: args.subdir || "",
  });
}

// 从锁定画布回传的完整工作流直接提交生成
export function submitGraph(workflow: unknown, url: string) {
  return apiPost<SubmitResult>("/comfyui/submit_graph", { workflow, url, client_id: comfyClientId() });
}

// 强行停止 ComfyUI 生图（人工打断工作流）：删排队项 + 中断执行。prompt_id 可空。
export function interruptComfy(url: string, promptId = "") {
  return apiPost<{ ok: boolean; deleted: boolean; interrupted: boolean }>(
    "/comfyui/interrupt",
    { url, prompt_id: promptId },
  );
}

File: backend/app/routers/comfyui.py
from fastapi import APIRouter, File, Form, HTTPException, Request, UploadFile
from pydantic import BaseModel

from app.config import COMFYUI_BASE_URL
from app.services import (
    comfyui_client, comfy_launcher, generation_store, image_store,
    local_media, workflow_submission,
)
from app.services.url_guard import validate_comfyui_url
from app.services.comfyui_client import ComfyError
from app.services.comfy_launcher import LaunchError

router = APIRouter()

_is_up = comfyui_client.is_up


@router.get("/")
def list_comfyui() -> dict[str, object]:
    return {"items": []}


class StartRequest(BaseModel):
    path: str          # ComfyUI 目录（含 main.py）
    url: str = COMFYUI_BASE_URL
    python_path: str = ""


@router.get("/status")
def status(url: str = COMFYUI_BASE_URL) -> dict[str, object]:
    return {"running": _is_up(url), "managed": comfy_launcher.is_managed()}


class ComfyConfig(BaseModel):
    path: str = ""
    url: str = COMFYUI_BASE_URL
    python_path: str = ""


@router.get("/config")
def get_config() -> ComfyConfig:
    """读 ComfyUI 路径/地址配置（供 start-dev 脚本等读取）。"""
    return ComfyConfig(**comfy_launcher.load_config())


@router.post("/config")
def set_config(cfg: ComfyConfig) -> ComfyConfig:
    """保存 ComfyUI 路径/地址，落盘到 data/comfy_config.json（ps1 脚本据此启动）。"""
    return ComfyConfig(**comfy_launcher.save_config(cfg.path, cfg.url, cfg.python_path))


@router.post("/start")
def start(req: StartRequest) -> dict[str, object]:
    try:
        return comfy_launcher.start(req.path, req.url, req.python_path)
    except LaunchError as e:
        raise HTTPException(status_code=e.status, detail=e.detail)


@router.post("/stop")
def stop(req: StartRequest) -> dict[str, object]:
    """关闭 ComfyUI（装插件/依赖后需先关）。"""
    return comfy_launcher.stop(req.url)


@router.post("/restart")
def restart(req: StartRequest) -> dict[str, object]:
    """重启 ComfyUI：先关再起（装完插件生效）。需提供 path 以重新拉起。"""
    try:
        return comfy_launcher.restart(req.path, req.url, req.python_path)
    except LaunchError as e:
        raise HTTPException(status_code=e.status, detail=e.detail)


class SubmitRequest(BaseModel):
    template_id: str
    values: dict[str, object] = {}   # key = "node_id.field" -> 覆盖值
    prompt: str = ""                 # 可选：自动注入到模板 prompt_node_id 的文本字段（/g 用）
    url: str = COMFYUI_BASE_URL
    client_id: str = ""              # 前端 WebSocket clientId，回传给 ComfyUI 定向推进度


@router.post("/submit")
def submit(req: SubmitRequest) -> dict[str, object]:
    """/s 启动：读模板原始工作流 → 转 API 格式 → 套用用户填的值 → 提交 /prompt。"""
    try:
        return workflow_submission.submit_template(
            req.template_id, req.values, req.prompt, req.url, req.client_id,
        )
    except workflow_submission.WorkflowSubmissionError as exc:
        raise HTTPException(status_code=exc.status, detail=exc.detail)


class SubmitGraphRequest(BaseModel):
    workflow: dict[str, object]      # iframe 回传的完整工作流 JSON（含用户改过的 widget 值）
    url: str = COMFYUI_BASE_URL
    client_id: str = ""              # 前端 WebSocket clientId，回传给 ComfyUI 定向推进度


@router.post("/submit_graph")
def submit_graph(req: SubmitGraphRequest) -> dict[str, object]:
    """从锁定画布回传的完整工作流直接转 API 并提交（用户在真实节点里改的值都在里面）。"""
    try:
        return workflow_submission.submit_graph(req.workflow, req.url, req.client_id)
    except workflow_submission.WorkflowSubmissionError as exc:
        raise HTTPException(status_code=exc.status, detail=exc.detail)


@router.post("/upload")
def upload_image(
    file: UploadFile = File(...),
    url: str = Form(COMFYUI_BASE_URL),
) -> dict[str, object]:
    """把图片转发上传到 ComfyUI 的 input 目录，返回其文件名（供 LoadImage 引用）。"""
    try:
        url = validate_comfyui_url(url)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    if not _is_up(url):
        raise HTTPException(status_code=400, detail="ComfyUI 未运行，无法上传图片")
    try:
        data = file.file.read()
        ref = comfyui_client.upload_image(url, file.filename, data, file.content_type or "image/png")
    except ComfyError as e:
        raise HTTPException(status_code=e.status, detail=e.detail)
    return {"name": ref}


@router.get("/result")
def result(prompt_id: str, url: str = COMFYUI_BASE_URL, node_ids: str = "") -> dict[str, object]:
    """轮询某次生成的状态与产出图片。
    node_ids 逗号分隔时，只保留这些节点的产物（主输出节点过滤）。
    """
    filter_ids = [n.strip() for n in node_ids.split(",") if n.strip()] if node_ids else None
    try:
        url = validate_comfyui_url(url)
        return comfyui_client.fetch_result(url, prompt_id, filter_ids)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except ComfyError as e:
        raise HTTPException(status_code=e.status, detail=e.detail)


class FinalizeGenerationImage(BaseModel):
    filename: str
    subfolder: str = ""
    type: str = "output"


class FinalizeGenerationRequest(BaseModel):
    thread_id: str
    repo_id: str
    prompt_id: str
    prompt: str = ""
    images: list[FinalizeGenerationImage] = []
    videos: list[FinalizeGenerationImage] = []
    output_dir: str = ""
    comfyui_url: str = COMFYUI_BASE_URL
    embed_base: str = ""
    embed_key: str = ""
    embed_model: str = "text-embedding-3-small"
    chat_base: str = ""
    chat_key: str = ""
    chat_model: str = ""
    regeneration: dict | None = None


@router.post("/finalize-generation")
def finalize_generation(req: FinalizeGenerationRequest) -> dict[str, object]:
    """持久化一批已完成的工作流产出；查询结果的 GET 路由保持无副作用。"""
    try:
        return generation_store.finalize_workflow_batch(
            thread_id=req.thread_id,
            repo_id=req.repo_id,
            prompt_id=req.prompt_id,
            prompt=req.prompt,
            images=[image.model_dump() for image in req.images],
            videos=[video.model_dump() for video in req.videos],
            output_dir=req.output_dir,
            comfyui_url=req.comfyui_url,
            embed_base=req.embed_base,
            embed_key=req.embed_key,
            embed_model=req.embed_model,
            chat_base=req.chat_base,
            chat_key=req.chat_key,
            chat_model=req.chat_model,
            regeneration=req.regeneration,
        )
    except ValueError as exc:
        raise HTTPException(status_code=400, detail=str(exc))


class InterruptRequest(BaseModel):
    url: str = COMFYUI_BASE_URL
    prompt_id: str = ""            # 有则先从队列删除未执行项，再中断正在执行的


@router.post("/interrupt")
def interrupt(req: InterruptRequest) -> dict[str, object]:
    """强行停止 ComfyUI 生图（人工打断工作流用）。

    prompt_id 有值：先 POST /queue {"delete":[id]} 删掉排队中未执行项，
    再 POST /interrupt 中断正在执行的任务（覆盖排队/执行两种状态）。
    容错：ComfyUI 未起/已完成都不报错，返回 ok。
    """
    try:
        req.url = validate_comfyui_url(req.url)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    res = comfyui_client.interrupt(req.url, req.prompt_id)
    return {"ok": True, **res}


@router.get("/view")
def view(filename: str, type: str = "output", subfolder: str = "", url: str = COMFYUI_BASE_URL):
    """代理 ComfyUI 的 /view，返回图片二进制（避免前端跨域直连 8188）。"""
    from fastapi.responses import Response

    try:
        url = validate_comfyui_url(url)
        data, ctype = comfyui_client.fetch_view(url, filename, type, subfolder)
    except ValueError as e:
        raise HTTPException(status_code=400, detail=str(e))
    except ComfyError as e:
        raise HTTPException(status_code=e.status, detail=e.detail)
    return Response(content=data, media_type=ctype)


class SaveLocalRequest(BaseModel):
    filename: str = ""                   # ComfyUI 产物文件名（ComfyUI 模式）
    subfolder: str = ""
    type: str = "output"
    repo_id: str = "home"
    output_dir: str = ""                 # 设置里的输出图片路径
    url: str = COMFYUI_BASE_URL
    src: str = ""                        # 通用模式：完整图片 URL 或 data URI（云端生图用）
    subdir: str = ""                     # 落到 <repo>/<subdir>/ 子夹（用户上传参考图 → reference）


@router.post("/save-local")
def save_local(req: SaveLocalRequest) -> dict[str, object]:
    """把原图存到设置的 outputDir（全分辨率，不降质），返回本地访问 URL。

    两种来源：
    - ComfyUI 模式：给 filename，从 ComfyUI /view 取原图。
    - 通用模式：给 src（http(s) URL 或 data:image/...;base64,...），直接下载/解码（云端生图）。
    ComfyUI 未起/清理 output 后仍可显示与「再改进」，且不依赖在线。
    """
    try:
        path = image_store.save_local(
            req.output_dir,
            req.repo_id,
            src=req.src,
            filename=req.filename,
            subfolder=req.subfolder,
            type=req.type,
            url=validate_comfyui_url(req.url),
            subdir=req.subdir,
        )
    except ComfyError as e:
        raise HTTPException(status_code=e.status, detail=e.detail)
    return {"ok": True, "path": path}


@router.get("/local-view")
def local_view(path: str, request: Request):
    """读取已留存到 outputDir 的本地文件，支持 Range 请求（视频拖动进度需要）。

    安全：只服务媒体文件（图片/视频）。该端点按设计要读任意本地路径（对话背景图允许
    用户填任意图片完整路径），无法按目录 jail，故用扩展名白名单挡住读取 .env/.db/源码等
    敏感文件的 LFI 攻击面。
    """
    try:
        media = local_media.open_local_media(path, request.headers.get("Range"))
    except local_media.LocalMediaError as exc:
        raise HTTPException(status_code=exc.status, detail=exc.detail, headers=exc.headers)
    if not media.partial:
        from fastapi.responses import FileResponse
        return FileResponse(media.path, media_type=media.media_type, headers=media.headers)
    from fastapi.responses import StreamingResponse
    return StreamingResponse(
        media.iter_bytes(), status_code=206, media_type=media.media_type, headers=media.headers,
    )
